File: 2026/AHC059/a05.cpp
#include <bits/stdc++.h>
using namespace std;

static constexpr int N = 20;
static constexpr double TIME_LIMIT = 1.9;
static constexpr int MAX_WIDTH = 1000;

struct Pos {
    int x, y;
};

static inline int manhattan(const Pos& a, const Pos& b) {
    return abs(a.x - b.x) + abs(a.y - b.y);
}

struct IsUsedBB {
    std::bitset<N * N> board;
    inline bool is_used(int x, int y) const { return board.test(x * N + y); }
    inline void set_used(int x, int y) { board.set(x * N + y); }
};

static inline Pos get_pair_pos(int num, const vector<array<Pos, 2>>& idx, const Pos& other) {
    const Pos& p1 = idx[num][0];
    const Pos& p2 = idx[num][1];
    if (p1.x == other.x && p1.y == other.y) return p2;
    return p1;
}

struct Node {
    IsUsedBB used;
    Pos current_pos{0, 0};
    bool has_stack_top = false;
    Pos stack_top{0, 0};

    int prev_path_length = 0;
    const Node* prev_node = nullptr;

    // Bounded container の lazy deletion 用
    bool alive = false;

    inline int total_cost() const {
        return has_stack_top ? (prev_path_length + manhattan(current_pos, stack_top)) : prev_path_length;
    }
};

// 幅制限付きの「best pop（min）」＋「worst drop（max）」両対応コンテナ
class BoundedBeam {
public:
    explicit BoundedBeam(int limit) : limit_(limit) {}

    bool empty() const { return alive_cnt_ == 0; }
    int alive_size() const { return alive_cnt_; }

    // いまの worst cost（alive がいなければ INF）
    int worst_cost() {
        clean_max();
        if (alive_cnt_ == 0) return INF;
        return maxpq_.top().cost;
    }

    // cost の候補が入る見込みがあるか（入らないならノード確保自体を避ける）
    bool would_accept(int cost) {
        clean_max();
        if (alive_cnt_ < limit_) return true;
        return cost < maxpq_.top().cost;
    }

    // ノードを追加（必要なら worst を落として幅を維持）
    void insert(Node* node) {
        node->alive = true;
        int cost = node->prev_path_length;
        minpq_.push({cost, node});
        maxpq_.push({cost, node});
        alive_cnt_++;

        if (alive_cnt_ > limit_) remove_worst();

        // lazy 削除のゴミが増えすぎると PQ 自体のメモリが増えるので適宜再構築
        maybe_rebuild();
    }

    // best（最小 cost）を取り出す（alive から外れる）
    Node* pop_best() {
        while (true) {
            clean_min();
            if (minpq_.empty()) return nullptr; // alive_cnt_==0 のはずだが保険
            auto e = minpq_.top(); minpq_.pop();
            if (!e.node->alive) continue; // stale
            e.node->alive = false;
            alive_cnt_--;
            return e.node;
        }
    }

    // コンテナ内の alive ノードを全て回収（探索終了後の最終選択用）
    vector<Node*> extract_all_alive() {
        vector<Node*> res;
        res.reserve(alive_cnt_);
        clean_min();
        while (!minpq_.empty()) {
            auto e = minpq_.top(); minpq_.pop();
            if (e.node->alive) {
                e.node->alive = false;
                res.push_back(e.node);
            }
        }
        // 状態クリア
        alive_cnt_ = 0;
        while (!maxpq_.empty()) maxpq_.pop();
        return res;
    }

private:
    static constexpr int INF = 1e9;

    struct Entry {
        int cost;
        Node* node;
    };
    struct MinCmp {
        bool operator()(const Entry& a, const Entry& b) const { return a.cost > b.cost; } // min-heap
    };
    struct MaxCmp {
        bool operator()(const Entry& a, const Entry& b) const { return a.cost < b.cost; } // max-heap
    };

    int limit_;
    int alive_cnt_ = 0;

    priority_queue<Entry, vector<Entry>, MinCmp> minpq_;
    priority_queue<Entry, vector<Entry>, MaxCmp> maxpq_;

    void clean_min() {
        while (!minpq_.empty() && !minpq_.top().node->alive) minpq_.pop();
    }
    void clean_max() {
        while (!maxpq_.empty() && !maxpq_.top().node->alive) maxpq_.pop();
    }

    void remove_worst() {
        while (true) {
            clean_max();
            if (maxpq_.empty()) return; // 保険
            auto e = maxpq_.top(); maxpq_.pop();
            if (!e.node->alive) continue;
            e.node->alive = false;
            alive_cnt_--;
            return;
        }
    }

    void maybe_rebuild() {
        // stale が多いと PQ が膨らむので、サイズが上限の数倍になったら掃除
        // （閾値は適当。必要に応じて調整可能）
        if ((int)minpq_.size() <= 6 * max(1, alive_cnt_) &&
            (int)maxpq_.size() <= 6 * max(1, alive_cnt_)) return;

        // minpq から alive を集めて再構築
        vector<Node*> alive_nodes;
        alive_nodes.reserve(alive_cnt_);

        clean_min();
        while (!minpq_.empty()) {
            auto e = minpq_.top(); minpq_.pop();
            if (e.node->alive) alive_nodes.push_back(e.node);
        }
        while (!maxpq_.empty()) maxpq_.pop();

        // 再投入
        for (Node* nd : alive_nodes) {
            int cost = nd->prev_path_length;
            minpq_.push({cost, nd});
            maxpq_.push({cost, nd});
        }
        // alive_cnt_ は変化なし
    }
};

static vector<char> make_commands(const vector<Pos>& collect_order, Pos current_pos) {
    vector<char> commands;
    int x = current_pos.x, y = current_pos.y;

    for (const auto& target : collect_order) {
        int tx = target.x, ty = target.y;

        while (x < tx) { commands.push_back('D'); x++; }
        while (x > tx) { commands.push_back('U'); x--; }
        while (y < ty) { commands.push_back('R'); y++; }
        while (y > ty) { commands.push_back('L'); y--; }

        commands.push_back('Z');
    }
    return commands;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int Nin;
    cin >> Nin; // 固定なので無視

    vector<vector<int>> grid(N, vector<int>(N));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) cin >> grid[i][j];
    }

    const int M = (N * N) / 2; // 200
    vector<array<Pos, 2>> nums_idx_list(M);
    vector<int> cnt(M, 0);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int num = grid[i][j];
            int k = cnt[num]++;
            nums_idx_list[num][k] = Pos{i, j};
        }
    }

    // ノードは採用されたものだけ確保する（ここがメモリ対策の要点）
    deque<Node> pool;

    pool.emplace_back();
    Node* root = &pool.back();
    root->used = IsUsedBB{};
    root->current_pos = Pos{0, 0};
    root->has_stack_top = false;
    root->prev_path_length = 0;
    root->prev_node = nullptr;

    const int LEVELS = M + 1; // 201
    vector<BoundedBeam> levels;
    levels.reserve(LEVELS);
    for (int i = 0; i < LEVELS; i++) levels.emplace_back(MAX_WIDTH);

    levels[0].insert(root);

    auto start = chrono::steady_clock::now();
    bool flag = true;

    while (flag) {
        for (int lv = 0; lv < M; lv++) {
            if (levels[lv].empty()) continue;

            Node* node = levels[lv].pop_best();
            if (!node) continue;

            // node から lv+1 を生成
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (node->used.is_used(i, j)) continue;

                    Pos cand{i, j};
                    int num = grid[i][j];
                    Pos pair_pos = get_pair_pos(num, nums_idx_list, cand);

                    int dist1 = manhattan(node->current_pos, cand);
                    int dist2 = node->has_stack_top ? manhattan(node->stack_top, pair_pos) : 0;
                    int new_cost = node->prev_path_length + dist1 + dist2;

                    // ここで「入らない候補は確保しない」ことでメモリを劇的に削減
                    if (!levels[lv + 1].would_accept(new_cost)) continue;

                    pool.emplace_back();
                    Node* nn = &pool.back();
                    nn->used = node->used;
                    nn->used.set_used(i, j);
                    nn->used.set_used(pair_pos.x, pair_pos.y);

                    nn->current_pos = cand;
                    nn->has_stack_top = true;
                    nn->stack_top = pair_pos;
                    nn->prev_path_length = new_cost;
                    nn->prev_node = node;

                    levels[lv + 1].insert(nn);
                }
            }

            double elapsed = chrono::duration<double>(chrono::steady_clock::now() - start).count();
            if (elapsed > TIME_LIMIT) { flag = false; break; }
        }
    }

    // Python版は最終レベルで total_cost 最小を選ぶ。
    // ただしタイムアウトで最終レベルが空の可能性があるので、最深の非空レベルから選ぶ。
    Node* best_node = root;
    for (int lv = LEVELS - 1; lv >= 0; lv--) {
        if (!levels[lv].empty()) {
            auto nodes = levels[lv].extract_all_alive();
            best_node = nodes[0];
            for (Node* nd : nodes) {
                if (nd->total_cost() < best_node->total_cost()) best_node = nd;
            }
            break;
        }
    }

    // 経路再構築（Python版の挙動を踏襲：後半 path_2 は reverse しない）
    vector<Pos> path;
    vector<Pos> path_2;
    const Node* cur = best_node;
    while (cur != nullptr) {
        path.push_back(cur->current_pos);
        if (cur->has_stack_top) path_2.push_back(cur->stack_top);
        cur = cur->prev_node;
    }
    if (!path.empty()) path.pop_back(); // (0,0) 期待
    reverse(path.begin(), path.end());
    path.insert(path.end(), path_2.begin(), path_2.end());

    auto commands = make_commands(path, Pos{0, 0});
    for (char c : commands) cout << c << "\n";

    return 0;
}

File: 2026/AHC059/a05.py
import os
from typing import Tuple, List, Optional
from collections import deque
import sys
import time
import heapq

MyPC = os.path.basename(__file__) != "Main.py"
if MyPC:
  from icecream import ic
  ic.disable()
else:
  def ic(*args):
    return None

ic.enable() if MyPC else None

N = 20  # グリッドのサイズ。20固定
TIME_LIMIT = 1.9  # 秒数制限
MAX_WIDTH = 1000  # chokudai_levelの最大幅

start_time = time.time()  # 開始時刻を記録

class IsUsedBB:
  """ ビットボードで使用済みマスを管理するクラス """
  def __init__(self):
    self.board = 0  # 20x20のビットボードを初期化

  def is_used(self, x: int, y: int) -> bool:
    return (self.board >> (x * N + y)) & 1 == 1

  def set_used(self, x: int, y: int):
    self.board |= (1 << (x * N + y))

  def copy(self) -> 'IsUsedBB':
    new_bb = IsUsedBB()
    new_bb.board = self.board  # intなのでそのままコピーでOK
    return new_bb

def is_valid_pos(x: int, y: int) -> bool:
  return 0 <= x < N and 0 <= y < N

def manhattan_dist(pos1: Tuple[int, int], pos2: Tuple[int, int]) -> int:
  """ マンハッタン距離を計算する """
  return abs(pos1[0] - pos2[0]) + abs(pos1[1] - pos2[1])

def nearest_valid(pos: Tuple[int, int], used: IsUsedBB) -> Tuple[int, int]:
  """ 最寄りの未使用マスを探す """
  x, y = pos
  min_dist = float("inf")

  # BFSで最寄りの未使用マスを探索
  queue = deque()
  queue.append((x, y, 0))  # (x座標, y座標, 距離)
  visited = [[False] * N for _ in range(N)]
  visited[x][y] = True
  directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
  while queue:
    cx, cy, dist = queue.popleft()

    if not used.is_used(cx, cy):
      return (cx, cy)

    for dx, dy in directions:
      nx, ny = cx + dx, cy + dy
      if is_valid_pos(nx, ny) and not visited[nx][ny]:
        visited[nx][ny] = True
        queue.append((nx, ny, dist + 1))

  # 見つからなかった場合
  raise ValueError("No unused cell found")

def get_pair_pos(num: int, nums_idx_list: List[List[Tuple[int, int]]], other_pos: Tuple[int, int]) -> Tuple[int, int]:
  """
  numのもう一方の位置を返す
  入力:
    num: 対応する数字
    nums_idx_list: 各数字の位置を格納するリスト
    other_pos: 既に知っている位置
  出力:
    numのもう一方の位置
  """
  pos1, pos2 = nums_idx_list[num]
  return pos2 if pos1 == other_pos else pos1

def make_commands(collect_order: List[Tuple[int, int]], current_pos: Tuple[int, int]) -> List[str]:
  """
  収集順序に基づいてコマンド列を生成する
  入力:
    collect_order: 収集順序のリスト
    current_pos: 現在位置
  出力:
    コマンド列のリスト
  """

  commands = []
  x, y = current_pos

  for target in collect_order:
    tx, ty = target
    
    # 縦移動
    while x < tx:
      commands.append('D')
      x += 1
    while x > tx:
      commands.append('U')
      x -= 1
    # 横移動
    while y < ty:
      commands.append('R')
      y += 1
    while y > ty:
      commands.append('L')
      y -= 1
    commands.append('Z')  # 収集コマンド
    x, y = tx, ty

  return commands

def get_path_length(path: List[Tuple[int, int]]) -> int:
  """ 経路の長さを計算する """
  length = 0
  for i in range(1, len(path)):
    length += manhattan_dist(path[i-1], path[i])
  return length

def find_best_next_pos(
  collect_order: List[Tuple[int, int]],
  collect_stack: List[Tuple[int, int]],
  used: IsUsedBB,
  nums_idx_list: List[List[Tuple[int, int]]],
  grid: List[List[int]],
) -> Tuple[int, int]:
  """
  次に収集するマスを決定する
  colect_ordrerの末尾と今回選択するマスの距離 + collect_stackの末尾と今回選択するマスの対のマスの距離 が最小になるマスを選ぶ
  入力:
    collect_order: 収集順序のリスト
    collect_stack: 2つめの数字を回収する際の経路のスタック
    used: 使用済みマスの管理リスト
    nums_idx_list: 各数字の位置を格納するリスト
    grid: グリッドの数字配置
  出力:
    次に収集するマスの位置
  """
  min_total_dist = float("inf")
  best_pos: Optional[Tuple[int, int]] = None

  for i in range(N):
    for j in range(N):
      if used.is_used(i, j):
        continue
      candidate_pos = (i, j)
      num = grid[i][j]
      pair_pos = get_pair_pos(num, nums_idx_list, candidate_pos)
      dist_1 = manhattan_dist(collect_order[-1], candidate_pos) if collect_order else 0  # collect_orderが空の場合は0
      dist_2 = manhattan_dist(collect_stack[-1], pair_pos) if collect_stack else 0  # collect_stackが空の場合は0
      total_dist = dist_1 + dist_2
      if total_dist < min_total_dist:
        min_total_dist = total_dist
        best_pos = candidate_pos

  if best_pos is None:
    raise ValueError("No unused cell found")
  return best_pos

# chokudaiサーチのノード
class Node:
  def __init__(self, used: IsUsedBB, current_pos: Tuple[int, int], stack_top: Optional[Tuple[int, int]] = None, prev_path_length: int = 0, prev_node: Optional['Node'] = None):
    """
    入力:
      used: 使用済みマスのビットボード
      current_pos: 現在位置
      stack_top: スタックのトップにある座標 (デフォルトは None)
      prev_path_length: これまでの経路長 (デフォルトは 0)
      prev_node: 前のノード (デフォルトは None)
    出力:      なし
    """
    self.used = used
    self.current_pos = current_pos
    self.stack_top = stack_top
    self.prev_path_length = prev_path_length
    self.prev_node = prev_node

  def __lt__(self, other: 'Node') -> bool:
    """ ノードの比較関数 (経路長で比較) """
    return self.prev_path_length < other.prev_path_length

  def next_nodes(self, grid: List[List[int]], nums_idx_list: List[List[Tuple[int, int]]]) -> List['Node']:
    """
    次のノードのリストを生成する
    入力:
      grid: グリッドの数字配置
      nums_idx_list: 各数字の位置を格納するリスト
    出力:
      次のノードのリスト
    """
    next_nodes = []
    for i in range(N):
      for j in range(N):
        if self.used.is_used(i, j):
          continue
        candidate_pos = (i, j)
        num = grid[i][j]
        pair_pos = get_pair_pos(num, nums_idx_list, candidate_pos)

        # 新しい使用済みビットボードを作成
        new_used = self.used.copy()
        new_used.set_used(i, j)
        new_used.set_used(pair_pos[0], pair_pos[1])

        # 新しい経路長を計算
        dist_1 = manhattan_dist(self.current_pos, candidate_pos)
        dist_2 = manhattan_dist(self.stack_top, pair_pos) if self.stack_top else 0
        new_path_length = self.prev_path_length + dist_1 + dist_2

        # 新しいノードを作成
        new_node = Node(
          used=new_used,
          current_pos=candidate_pos,
          stack_top=pair_pos,
          prev_path_length=new_path_length,
          prev_node=self
        )
        next_nodes.append(new_node)

    return next_nodes
  
  def reconstruct_path(self) -> List[Tuple[int, int]]:
    """
    ノードから経路を再構築する
    入力:      なし
    出力:
      経路のリスト
    """
    path = []
    path_2 = []  # スタックの方の経路を一時的に保存するリスト
    node = self
    while node is not None:
      path.append(node.current_pos)
      if node.stack_top is not None:
        path_2.append(node.stack_top)
      node = node.prev_node
    # 最初の座標(0,0)は重複しているので削除
    first = path.pop()  # ここでは末尾が最初の座標になる
    if not (first == (0, 0)):  # 安全確認
      raise ValueError("The first position is not (0,0)")
    path.reverse()  # 順方向に直す
    path.extend(path_2)  # スタックの方は最初から順方向なのでそのまま追加
    return path
  
  def total_cost(self) -> int:
    """
    ノードの総コストを計算する
    今までの経路長 + 現在位置からstackのトップまでの距離
    入力:      なし
    出力:
      総コスト
    """
    return self.prev_path_length + manhattan_dist(self.current_pos, self.stack_top) if self.stack_top else self.prev_path_length  # stack_topがNoneのときにこれが実行されることはないはずだが、念のため

def main():
  _ = int(input())  # Nは20で固定なので無視

  grid = [list(map(int, input().split())) for _ in range(N)]

  nums_idx_list = [[] for _ in range(N**2)]  # 各数字の位置を格納するリスト (各数字が2つずつ存在する)
  for i in range(N):
    for j in range(N):
      num = grid[i][j]
      nums_idx_list[num].append((i, j))
  
  used = IsUsedBB()  # 使用済みマスの管理

  # collect_order: List[Tuple[int, int]] = []  # 収集順序を格納するリスト

  current_pos = (0, 0)  # 現在位置

  # collect_stack = []  # 2つめの数字を回収する際の経路のスタック

# chokudaiサーチで収集順序を決定
  root_node = Node(used=used, current_pos=current_pos)
  chokudai_levels: List[List[Node]] = [[] for _ in range(N**2//2 + 1)]  # 各レベルのノードリスト。レベルごとにheapとして管理
  chokudai_levels[0].append(root_node)

  # for i in range(N**2//2):
  #   next_beam_nodes = []
  #   for node in beam_nodes:
  #     next_nodes = node.next_nodes(grid, nums_idx_list)
  #     next_beam_nodes.extend(next_nodes)
  #   # 経路長でソートして上位K個を残す
  #   if i < N**2//2 - 1:  # 最後のステップ以外では経路長でソート
  #     next_beam_nodes.sort(key=lambda n: n.prev_path_length)
  #   else:  # 最後のステップでは総コストでソート
  #     next_beam_nodes.sort(key=lambda n: n.total_cost())
  #   beam_nodes = next_beam_nodes[:K]

  flag = True

  while flag:
    for i in range(N**2//2):
      chokudai_level = chokudai_levels[i]
      next_chokudai_level = chokudai_levels[i + 1]
      if not chokudai_level:  # ノードが空ならスキップ
        continue
      node = heapq.heappop(chokudai_level)
      next_nodes = node.next_nodes(grid, nums_idx_list)
      for next_node in next_nodes:
        heapq.heappush(next_chokudai_level, next_node)
      # chokudai_levelの最大幅を制限
      while len(next_chokudai_level) > MAX_WIDTH:
        next_chokudai_level.pop()  # コストの高いノードをリストの末尾から通常の削除
      # 時間制限チェック
      elapsed_time = time.time() - start_time
      if elapsed_time > TIME_LIMIT:
        flag = False
        break

  # 最終レベルで総コストでソート
  chokudai_levels[-1].sort(key=lambda n: n.total_cost())

  # 最良ノードを取得
  best_node = chokudai_levels[-1][0]
  collect_order = best_node.reconstruct_path()

  # 経路を出力
  commands = make_commands(collect_order, (0, 0))
  for cmd in commands:
    print(cmd)

  # 総コストの確認 (デバッグ用)
  total_length = get_path_length(collect_order)
  print(f"Total path length: {total_length}", file=sys.stderr)

if __name__ == "__main__":
  main()